Feature: add proxmox_node_network and proxmox_node_network_info modules

aleskxyz opened the pull request:
##### SUMMARY
Add comprehensive network interface management capabilities to the Proxmox collection with two new modules: `proxmox_node_network` for managing network interfaces and `proxmox_node_network_info` for retrieving network interface information. These modules support all major Proxmox network interface types including bridges, bonds, VLANs, and Open vSwitch interfaces with staged configuration management to prevent accidental network disruption.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
proxmox_node_network, proxmox_node_network_info

##### ADDITIONAL INFORMATION
**New Modules Added:**
- `proxmox_node_network`: Manage network interfaces (bridge, bond, eth, vlan, OVS)
- `proxmox_node_network_info`: Retrieve network interface information

**Key Features:**
- Staged configuration management with apply/revert functionality
- Support for all major interface types (bridge, bond, eth, vlan, OVSBridge, OVSBond, OVSIntPort)
- Comprehensive parameter validation for each interface type
- Support for parameter deletion (string params with empty string, integer params with -1)
- Full check mode support
- Full diff mode support

Thulium-Drake commented:
@gyptazy Would you have time to review this? I'm not well-versed enough in Python to review this :-(

gyptazy commented:
> @gyptazy Would you have time to review this? I'm not well-versed enough in Python to review this :-(

Will do but can't promise to do it this weekend. Probably on Monday.

gyptazy commented on plugins/modules/proxmox_node_network.py:
```diff
@@ -0,0 +1,1734 @@
+    """Validate IPv6 CIDR format."""
+    if not cidr:
+        return False
+    pattern = r"^[0-9a-fA-F:]+/\d{1,3}$"
```
That might not be strict enough. Maybe this fits better:
```
r'^([0-9A-Fa-f:]+::?[0-9A-Fa-f:]*)/(12[0-8]|1[01][0-9]|[0-9]{1,2})$'
```
Also here, maybe better to use an external lib like `ipaddress`; even I'd like to avoid it.

IamLunchbox replied:
Since `ipaddress` is part of the standard library, using it would outsource complexity and not include major (installation) overhead.

Out of curiosity: Why would you like avoid including this library otherwise?

gyptazy commented:
Hey @aleskxyz,

that's well written and structured - appreciate your work! There're just some smaller things that're not yet important (but might need some changes in future releases) but it doesn't make sense or brings any huge benefits to adjust this right now. I didn't give OVS a try but apart from that it looks good to me. Just the regex might fail. I mean, it would be a user error but when we already have a validate function, we should do a proper validation (I added a quick idea, could probably also be improved).

So far, well done!

cc: @Thulium-Drake 

Cheers,
gyptazy

aleskxyz commented:
> Hey @aleskxyz,
> 
> that's well written and structured - appreciate your work! There're just some smaller things that're not yet important (but might need some changes in future releases) but it doesn't make sense or brings any huge benefits to adjust this right now. I didn't give OVS a try but apart from that it looks good to me. Just the regex might fail. I mean, it would be a user error but when we already have a validate function, we should do a proper validation (I added a quick idea, could probably also be improved).
> 
> So far, well done!
> 
> cc: @Thulium-Drake
> 
> Cheers, gyptazy

Hi @gyptazy 
I have updated the code and refactored the logic with the help of the ipaddress module. I have also mentioned it as a requirement in the document section.
The feature branch has been rebased onto the latest commit on the main branch.
Thanks.

Thulium-Drake commented:
@aleskxyz Thanks for your work! 

@gyptazy @IamLunchbox Are you happy with the results? :-)

gyptazy commented:
> @aleskxyz Thanks for your work!

Also from me, thanks again for your work and sorry for the delay!

> @gyptazy @IamLunchbox Are you happy with the results? :-)

I’m currently abroad and will have a look at it next Wednesday, when I have a bigger screen again.

Thulium-Drake commented:
@gyptazy how would you like to proceed? Are the modules fit for use (so we can process the changes you requested in later PRs if needed?)?
